### PR TITLE
Ccg 81 create dummy target that can take damage

### PR DESCRIPTION
**Pull request by aaronspringer**
Changes
* Created Dummy that can take damage and die
* Created projectile that is instantiated when the player attacks
  * Projectile is destroyed when hitting an enemy, 1 pierce
  * There is currently no projectile maximum lifespan

### CODE REVIEW

File: scripts/entity/base_entity.gd [hunk 1]
```diff
@@ -15,8 +15,14 @@ class_name BaseEntity extends Node
 func _ready() -> void:
 
 	# if the movement component and input component exist connect the movement together.
-	input_component.on_horizontal_movement_input.connect(movement_component.on_movement_input)
+	if movement_component and input_component:
+		input_component.on_horizontal_movement_input.connect(movement_component.on_movement_input)
+		input_component.on_jump_input_started.connect(movement_component.on_jump_input_started)
+		input_component.on_jump_input.connect(movement_component.on_jump_input)
+		input_component.on_jump_input_cancelled.connect(movement_component.on_jump_input_cancelled)
```
Comment (SllimRevilo) [inline scripts/entity/base_entity.gd:22]: this should go away in a rebase. if statement was taken out

File: scripts/entity/components/attack_components/player_attack_component.gd [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+func _ready() -> void:
+	# Override the abstract class
+	pass
+
+func on_attack_input_started() -> void:
+	# Create an instance of the projectile
+	var projectile_instance: BaseProjectileComponent = projectile_scene.instantiate()
+	add_child(projectile_instance)
+	# Initialize the projectile
+	projectile_instance.init(projectile_spawn_position.global_position, 5, [1], 0, 200, Vector2.ZERO)
```
Comment (SllimRevilo) [inline scripts/entity/components/attack_components/player_attack_component.gd]: I think the values here will need to be exported but this should be fine for now. I think a couple architecture changes will be needed as well but again should be fine for this

File: scripts/entity/components/health_and_status_components/base_health_and_status_component.gd [hunk 3]
```diff
@@ -1,11 +1,72 @@
...
+	if _current_health > 0:
+		var final_health_gain: int = min((_current_health + health_to_heal + flat_heal_bonus) * percent_heal_bonus)
+		_current_health = min(final_health_gain, max_health)
+		on_healed.emit()
+
+
+## Checks the entity's health and emits death or health change signals.
+func _check_health_status() -> void:
+	if _current_health <= 0:
+		on_death.emit()
```
Comment (SllimRevilo) [inline scripts/entity/components/health_and_status_components/base_health_and_status_component.gd:55]: I would just move these to the one place this function is called and connect the die function to the on death signal in _ready

File: scripts/entity/components/input_components/player_input_component.gd [hunk 4]
```diff
@@ -5,6 +5,7 @@ class_name PlayerInputComponent extends BaseInputComponent
 const JUMP_INPUT_NAME = "jump"
 const LEFT_INPUT_NAME = "left"
 const RIGHT_INPUT_NAME = "right"
+const ATTACK_INPUT_NAME = "attack"
```
Comment (SllimRevilo) [inline scripts/entity/components/input_components/player_input_component.gd]: These all should be StringNames. The top three should be fixed after conflicts resolved

File: scripts/entity/components/health_and_status_components/base_health_and_status_component.gd [hunk 5]
```diff
@@ -1,11 +1,72 @@
...
+
+# func remove_status_effect(status_effect: StatusEffect):
+	# TBD
+
+
+## Reduces health based on the damage taken and the entity's defense.
+## [param damage_to_take] int: The raw damage taken.
+func take_damage(damage_to_take: int) -> void:
+	print("Taking " + str(damage_to_take))
+	if _current_health > 0:
```
Comment (SllimRevilo) [inline scripts/entity/components/health_and_status_components/base_health_and_status_component.gd]: why are we checking if the hp is > 0

File: scripts/entity/components/health_and_status_components/base_health_and_status_component.gd [hunk 6]
```diff
@@ -1,11 +1,72 @@
...
+		var final_damage = max((damage_to_take - flat_defense_bonus) * percent_defense_bonus, 0)
```
Comment (SllimRevilo) [inline scripts/entity/components/health_and_status_components/base_health_and_status_component.gd]: we can declare the type

File: scripts/entity/components/health_and_status_components/base_health_and_status_component.gd [hunk 7]
```diff
@@ -1,11 +1,68 @@
...
+	if _current_health <= 0:
+		on_death.emit()
+
+
+
 
+## Heals the entity, ensuring it doesn't go over max health.
+## [param health_to_heal] int: The amount of health to restore.
+func heal_damage(health_to_heal: int) -> void:
+	if _current_health > 0:
```
Comment (SllimRevilo) [inline scripts/entity/components/health_and_status_components/base_health_and_status_component.gd]: i dont think we need to check this

File: scripts/entity/components/health_and_status_components/base_health_and_status_component.gd [hunk 8]
```diff
@@ -1,11 +1,68 @@
...
+		var final_health_gain: int = min((_current_health + health_to_heal + flat_heal_bonus) * percent_heal_bonus)
+		_current_health = min(final_health_gain, max_health)
+		on_healed.emit()
 
-# Called every frame. 'delta' is the elapsed time since the previous frame.
-func _process(_delta: float) -> void:
-	pass
+## Free's the queue of the entity
```
Comment (SllimRevilo) [inline scripts/entity/components/health_and_status_components/base_health_and_status_component.gd]: destroys the parent (the entity) and all its children

File: scripts/entity/components/projectile_components/base_projectile_component.gd [hunk 9]
```diff
@@ -0,0 +1,36 @@
+class_name BaseProjectileComponent extends Area2D
+
+var damage: int = 20 # Damage inflicted by the projectile
+var healing: int = 0
+var move_speed: int = 100
+var on_hit_effects: Array[int] = [] # Integer for now, change to status_effect once implemented
+var target_location: Vector2 = Vector2.ZERO
+var velocity: Vector2 = Vector2.ZERO
```
Comment (SllimRevilo) [inline scripts/entity/components/projectile_components/base_projectile_component.gd]: If they have default values they all should just be zeroed out.

**Review by SllimRevilo — APPROVED** [review]

File: scripts/entity/components/attack_components/player_attack_component.gd [hunk 10]
```diff
@@ -0,0 +1,26 @@
+class_name PlayerAttackComponent extends BaseAttackComponent
```
Comment (Smoresies) [inline scripts/entity/components/attack_components/player_attack_component.gd]: Could be renamed "Ranged Attack Component"

**Review by Smoresies — APPROVED** [review]

**Pull request merged**